fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -114,26 +122,52 @@ public void beforeSearch(FHIRPersistenceEvent event) throws FHIRPersistenceInter
 
             if (compartment != null) {
                 // Compartment search - validate patient access
-                if (!"Patient".equals(compartment)) {
-                    String msg = "Compartment search with compartment '" + compartment + "' is not permitted.";
-                    throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
-                }
-                if (!patientIdFromToken.contains(compartmentId)) {
-                    String msg = "Interaction with 'Patient/" + compartmentId + "' is not permitted under patient context '" + patientIdFromToken + "'.";
+                switch(compartment) {
+                case "Patient":
+                    if (!patientIdFromToken.contains(compartmentId)) {
+                        String msg = "Interaction with 'Patient/" + compartmentId + "' is not permitted under patient context '" + patientIdFromToken + "'.";
+                        throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
+                    }
+                    try {
+                        // TODO: confirm this entire try-catch block is safe to remove
+                        // the REST layer should already prevent compartment searches for invalid resource types
+                        if (!CompartmentUtil.getCompartmentResourceTypes(PATIENT).contains(event.getFhirResourceType())) {
+                            String msg = "Resource type '" + event.getFhirResourceType() + "' is not valid for Patient compartment search.";
+                            throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.INVALID));
+                        }
+                    } catch (FHIRSearchException e) {
+                        log.log(Level.WARNING, "Unexpected exception while getting compartment resource types", e);
+                    }
+                    break;
+                case "Encounter":
+                    Resource r = executeRead(event.getPersistenceImpl(), ModelSupport.getResourceType(compartment), compartmentId);
+                    if (!isInCompartment(r, CompartmentType.PATIENT, patientIdFromToken)) {
+                        String msg = "Access to compartment 'Device/" + compartmentId + "' is not forbidden for patient context " + patientIdFromToken;
```

Comment:
```suggestion
                        String msg = "Access to compartment 'Encounter/" + compartmentId + "' is forbidden for patient context " + patientIdFromToken;
```

Reply:
I changed this to `"Interaction with 'Encounter/" + compartmentId + "' is not permitted for patient context " + patientIdFromToken`